File: Algorithm_DataStructures/Java/LinkedList/src/LinkedList.java
/**
 * Created by VarunPius on 12/28/2015.
 */

class Node{
    int data;
    Node next;

    Node(int data)
    {
        this.data = data;
    }
}

class LinkedListFunc
{
    Node head;

    public void insertAtEnd(Node n)
    {
        if (head == null)
        {
            head = n;
        }
        else
        {
            Node tmp = head;
            while (tmp.next != null)
                tmp = tmp.next;

            tmp.next = n;
        }

    }

    public void printLinkedList(Node head)
    {
        Node tmp = head;

        while (tmp != null)
        {
            System.out.print(tmp.data + " -> ");
            tmp = tmp.next;
        }
    }

    public void reverseLinkedList(Node head)
    {
        Node curr = head;
        Node previousNode = null;
        Node nextNode = null;

        Node tmp = previousNode;
        nextNode = previousNode;
        curr.next = previousNode;
        nextNode.next = curr;
        nextNode.next.next = previousNode;

    }
}
public class LinkedList {
    public static void main(String[] args)
    {
        Node head = new Node(10);

        LinkedListFunc list = new LinkedListFunc();

        list.insertAtEnd(head);
        list.insertAtEnd(new Node(4));
        list.insertAtEnd(new Node(3));
        list.insertAtEnd(new Node(9));
        Node tail = new Node(11);
        list.insertAtEnd(tail);

        list.printLinkedList(head);
        System.out.println();
        list.printLinkedList(tail);
    }
}

class reverseLinkedList
{

}

File: Algorithm_DataStructures/Python/LinkedList/LinkedList.py
__author__ = 'VarunPius'


class Node:
    def __init__(self, data = None):
        self.data = data
        self.next = None

    def setNext(self, next_node):
        self.next = next_node

    def hasNext(self):
        return self.next is not None

    def getNext(self):
        return self.next

    def getData(self):
        return self.data


class LinkedListFunc:
    ##Node head
    def __init__(self, head):
        self.first = head

    def insert_at_end(self, next_node):
        tmp_node = self.first

        '''if (tmp_node):
            while tmp_node.hasNext():
                tmp_node = tmp_node.getNext()
        '''

        if(tmp_node):
            while tmp_node.hasNext():       #tmp_node.next can also b used
                tmp_node = tmp_node.getNext()
                #tmp_node = tmp_node.next

        tmp_node.setNext(next_node)

    def printList(self):
        tmp_node = self.first

        if (tmp_node):
            while tmp_node:
                print(str(tmp_node.getData()) + " -> ") #instead of tmp_node.getData() we can use tmp_node.data
                tmp_node = tmp_node.getNext()  #instead of tmp_node.getNext() we can use tmp_node.next
        #tmp_node.getData()

hd = Node(5)
ll = LinkedListFunc(hd)

#ll.insert_at_end(hd)
ll.insert_at_end(Node(9))
ll.insert_at_end(Node(10))
ll.insert_at_end(Node(4))

ll.printList()
